debug-tools/amd-dbgapi.cmake

```diff
@@ -0,0 +1,676 @@
+
+      # Enable libbacktrace in the sources.
+      target_compile_definitions(amd-dbgapi PRIVATE HAVE_BACKTRACE)
+      # Set up including the header in the compilation and linking against
+      # libbacktrace.
+      target_include_directories(amd-dbgapi PRIVATE ${BACKTRACE_INCLUDE_DIR})
+      target_link_libraries(amd-dbgapi PRIVATE ${BACKTRACE_LIB})
+    endif()
+  else()
+    message(STATUS "[libbacktrace check] libbacktrace.a not found. Skipping.")
+  endif()
+endif()
+
+option(API_TRACING "Enable API tracing" ON)
+if(API_TRACING)
+  target_compile_definitions(amd-dbgapi PRIVATE WITH_API_TRACING)
+endif()
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  target_sources(amd-dbgapi PRIVATE src/os_driver_kfd.cpp src/utils_linux.cpp)
+
+  set(CMAKE_REQUIRED_FLAGS_BKP "${CMAKE_REQUIRED_FLAGS}")
+  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -std=c++17")
+  # For some platforms (PPC64), linux/types.h uses L64, but this code assumes
+  # LL64 (i.e.  __u64 should be a long long unsigned, not long unsigned).
+  check_cxx_source_compiles(
+    "#include <linux/types.h>
+     #include <type_traits>
+     int main () { static_assert (std::is_same_v<long long unsigned, __u64>); return 0; }"
+     LINUX_TYPES_HAS_LL64)
+
+  if(NOT LINUX_TYPES_HAS_LL64)
+    # Check if using __SANE_USERSPACE_TYPES__ gives us LL64
+    check_cxx_source_compiles(
+      "#define __SANE_USERSPACE_TYPES__
+       #include <linux/types.h>
+       #include <type_traits>
+       int main () { static_assert (std::is_same_v<long long unsigned, __u64>); return 0; }"
+       LINUX_TYPES_LL64_NEEDS_SANE_USERSPACE_TYPES)
+    if(LINUX_TYPES_LL64_NEEDS_SANE_USERSPACE_TYPES)
+      target_compile_definitions(amd-dbgapi PRIVATE "__SANE_USERSPACE_TYPES__")
+    else()
+      message(FATAL_ERROR
+              "A LL64 configuration is required for the Linux backend")
+    endif()
+  endif()
+  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS_BKP}")
+elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+  target_sources(amd-dbgapi PRIVATE src/os_driver_kmd.cpp src/utils_windows.cpp)
+
+  target_include_directories(amd-dbgapi PRIVATE
+    src/windows/imported
+    ${D3DKMT_INCLUDE_PATH}
+    third_party/libdxg/include)
+
+  target_compile_definitions(amd-dbgapi
+    PRIVATE D3DKMDT_SPECIAL_MULTIPLATFORM_TOOL)
+
+  # Including libdxg raises some warnings (unknown pargmas) as well us
+  # "X too small to hold all values of Y".  The latter warning cannot be
+  # silenced, se we need to disable -Werror for this file.
+  set_source_files_properties(
+    src/os_driver_kmd.cpp
+    PROPERTIES
+      COMPILE_OPTIONS "-Wno-error;-Wno-unknown-pragmas")
+else()
+  message(FATAL_ERROR "Platform ${CMAKE_SYSTEM_NAME} is not supported")
+endif()
+
+find_package(amd_comgr REQUIRED CONFIG
+  PATHS
+    /opt/rocm/
+  PATH_SUFFIXES
+    lib/cmake/amd_comgr
+)
+MESSAGE(STATUS "Code Object Manager found at ${amd_comgr_DIR}.")
+
+option(REQUIRE_THREADS "Require linking against Threads" OFF)
```

Comment:
Move this up to the upper part of the CMake config so that all options are at a central place.